br/pkg/lightning/backend/external/writer.go

```diff
@@ -0,0 +1,331 @@
+		writeBatch:     make([]common.KvPair, 0, b.writeBatchSize),
+		currentSeq:     0,
+		filenamePrefix: filenamePrefix,
+		writerID:       writerID,
+		kvStore:        nil,
+		onClose:        b.onClose,
+		closed:         false,
+	}
+}
+
+// Writer is used to write data into external storage.
+type Writer struct {
+	ctx            context.Context
+	store          storage.ExternalStorage
+	writerID       int
+	currentSeq     int
+	filenamePrefix string
+
+	kvStore *KeyValueStore
+	rc      *rangePropertiesCollector
+
+	memSizeLimit uint64
+
+	kvBuffer   *membuf.Buffer
+	writeBatch []common.KvPair
+
+	onClose OnCloseFunc
+	closed  bool
+
+	// Statistic information per batch.
+	batchSize uint64
+
+	// Statistic information per writer.
+	minKey    tidbkv.Key
+	maxKey    tidbkv.Key
+	totalSize uint64
+}
+
+// AppendRows appends rows to the external storage.
+// Note that this method is NOT thread-safe.
+func (w *Writer) AppendRows(ctx context.Context, _ []string, rows encode.Rows) error {
+	kvs := kv.Rows2KvPairs(rows)
+	if len(kvs) == 0 {
+		return nil
+	}
+	for _, pair := range kvs {
+		w.batchSize += uint64(len(pair.Key) + len(pair.Val))
+		buf := w.kvBuffer.AllocBytes(len(pair.Key))
+		key := append(buf[:0], pair.Key...)
+		val := w.kvBuffer.AddBytes(pair.Val)
+		w.writeBatch = append(w.writeBatch, common.KvPair{Key: key, Val: val})
+		if w.batchSize >= w.memSizeLimit {
+			if err := w.flushKVs(ctx); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (w *Writer) IsSynced() bool {
+	return false
+}
+
+func (w *Writer) Close(ctx context.Context) (backend.ChunkFlushStatus, error) {
+	if w.closed {
+		return status(true), nil
+	}
+	w.closed = true
+	defer w.kvBuffer.Destroy()
+	err := w.flushKVs(ctx)
+	if err != nil {
+		return status(false), err
+	}
+
+	logutil.BgLogger().Info("close writer",
+		zap.Int("writerID", w.writerID),
+		zap.String("minKey", hex.EncodeToString(w.minKey)),
+		zap.String("maxKey", hex.EncodeToString(w.maxKey)))
+
+	w.writeBatch = nil
+
+	w.onClose(&WriterSummary{
+		WriterID:  w.writerID,
+		Seq:       w.currentSeq,
+		Min:       w.minKey,
+		Max:       w.maxKey,
+		TotalSize: w.totalSize,
+	})
+	return status(true), nil
+}
+
+func (w *Writer) recordMinMax(newMin, newMax tidbkv.Key, size uint64) {
+	if len(w.minKey) == 0 || newMin.Cmp(w.minKey) < 0 {
+		w.minKey = newMin.Clone()
+	}
+	if len(w.maxKey) == 0 || newMax.Cmp(w.maxKey) > 0 {
+		w.maxKey = newMax.Clone()
+	}
+	w.totalSize += size
+}
+
+type status bool
+
+func (s status) Flushed() bool {
+	return bool(s)
+}
+
+func (w *Writer) flushKVs(ctx context.Context) error {
+	if len(w.writeBatch) == 0 {
+		return nil
+	}
+
+	dataWriter, statWriter, err := w.createStorageWriter()
+	if err != nil {
+		return err
+	}
+
+	ts := time.Now()
+	var saveBytes uint64
+
+	defer func() {
+		w.currentSeq++
+		err := dataWriter.Close(w.ctx)
+		if err != nil {
+			logutil.BgLogger().Error("close data writer failed", zap.Error(err))
+		}
+		err = statWriter.Close(w.ctx)
```

Comment:
the two closing is not atomic. what if TiDB node crashes between them and how will we find the dirty files when scanning s3? Do we treat the creating of stat file as the commit point of writing files to s3?

Reply:
Maybe we should add the validation check when TiDB reading from s3.

Reply:
Or maybe the framework will choose another `filenamePrefix` for the writer, and only expose the successful `filenamePrefix`s to following reader?